src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,455 @@
+		return err
+	}); err != nil {
+		return err
+	}
+	for _, commitInfo := range commitInfos {
+		if err := cb(commitInfo); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// applies a callback, cb, to all the commits in a commit set. A commit set includes
+func (d *driver) inspectCommitSet(ctx context.Context, commitset *pfs.CommitSet, wait bool, cb func(*pfs.CommitInfo) error) error {
+	if !wait {
+		return d.inspectCommitSetImmediate(ctx, commitset, cb)
+	}
+	sent := map[string]struct{}{}
+	send := func(ci *pfs.CommitInfo) error {
+		if _, ok := sent[pfsdb.CommitKey(ci.Commit)]; ok {
+			return nil
+		}
+		sent[pfsdb.CommitKey(ci.Commit)] = struct{}{}
+		return cb(ci)
+
+	}
+	unfinishedCommits := make([]*pfs.Commit, 0)
+	// NOTE: before 2.5, a triggered commits would be included as part of the same commit set as the triggering commit set,
+	// so we would wait for all of the current commit set to finish, then check if new previously unknown commits are added due to triggers.
+	if err := d.inspectCommitSetImmediate(ctx, commitset, func(ci *pfs.CommitInfo) error {
+		if ci.Finished != nil {
+			return send(ci)
+		}
+		unfinishedCommits = append(unfinishedCommits, proto.Clone(ci.Commit).(*pfs.Commit))
+		return nil
+	}); err != nil {
+		return err
+	}
+	for _, uc := range unfinishedCommits {
+		// TODO: make a dedicated call just for the blocking part, inspectCommit is a little heavyweight?
+		ci, err := d.inspectCommit(ctx, uc, pfs.CommitState_FINISHED)
+		if err != nil {
+			return err
+		}
+		if err := send(ci); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *driver) listCommitSet(ctx context.Context, project *pfs.Project, cb func(*pfs.CommitSetInfo) error) error {
+	// Track the commitsets we've already processed
+	seen := map[string]struct{}{}
+	// Return commitsets by the newest commit in each set (which can be at a different
+	// timestamp due to triggers or deferred processing)
+	commitInfo := &pfs.CommitInfo{}
+	err := d.commits.ReadOnly(ctx).List(commitInfo, col.DefaultOptions(), func(string) error {
+		if commitInfo.GetCommit().GetBranch().GetRepo().GetProject().GetName() != project.GetName() {
+			return nil
+		}
+		if _, ok := seen[commitInfo.Commit.ID]; ok {
+			return nil
+		}
+		seen[commitInfo.Commit.ID] = struct{}{}
+		var commitInfos []*pfs.CommitInfo
+		err := d.inspectCommitSet(ctx, &pfs.CommitSet{ID: commitInfo.Commit.ID}, false, func(ci *pfs.CommitInfo) error {
+			commitInfos = append(commitInfos, ci)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		return cb(&pfs.CommitSetInfo{
+			CommitSet: client.NewCommitSet(commitInfo.Commit.ID),
+			Commits:   commitInfos,
+		})
+	})
+	return errors.EnsureStack(err)
+}
+
+// oldestAncestor returns the oldest ancestor commit of 'startCommit' leveraging the pool of known commits in the 'skipSet'
+func oldestAncestor(startCommit *pfs.CommitInfo, skipSet map[string]*pfs.CommitInfo) *pfs.Commit {
```

Comment:
nit, since this is used in `deleteCommits`, can we have it follow the definition of `deleteCommits`. It reads a bit better imo than being above it.